101-convolution.rst

```diff
@@ -0,0 +1,202 @@
+Approved by
+Approved on
+Estimated effort   4 weeks
+Implemented by
+Completion date
+Implemented in
+=================  ==================================
+
+Developed in consultation with:
+Eric Hunsberger <eric.hunsberger@appliedbrainresearch.com>
+Trevor Bekolay <trevor.bekolay@appliedbrainresearch.com>
+
+Context
+=======
+
+Right now Nengo's Connection system is based on the assumption of dense
+(all-to-all) connectivity between objects (Ensembles/Nodes).
+We want to generalize this to support other connectivity patterns,
+e.g. sparse/convolutional weights.
+Non-dense connectivity patterns are quite common in neural network
+research/applications, so it would be good to have a way to represent them with
+a consistent interface across the Nengo ecosystem.
+
+Proposal
+========
+
+The proposal is that we add a new type of object that can be passed as the
+``transform`` argument on a ``Connection``, which would represent a
+convolutional transformation.  E.g.::
+
+  nengo.Connection(a.neurons, b.neurons,
+                   transform=nengo.Conv2D(kernel_size=(3, 3), stride=(1, 1))
+
+In this proposal we will focus on the specific case of convolutional
+connectivity (although with the general view that this
+approach should be extendable to other, related use cases in the future).
+In addition, we only propose to support these transform types on
+connections where the pre object is a Node or Neuron object (i.e., not an
+Ensemble, meaning decoded connection).
+These connections also will not support online learning rules.
+See "open questions/extensions" below for more discussion on these
+restrictions.
+
+Details
+=======
+
+Some of the other options considered were creating a new connection object,
+e.g.::
+
+  nengo.Conv2DConnection(a.neurons, b.neurons, kernel_size=(3, 3),
+                         stride=(1, 1))
+
+or a ``Network`` that would encapsulate lower level objects implementing
+the convolutional connection, e.g.::
+
+  conv = nengo.networks.Conv2D(
+      a.neurons, b.neurons, kernel_size=(3, 3), stride=(1, 1))
+
+The ``transform``-based approach seemed the best option, as it adds minimal
+complexity to the Nengo interface, avoids duplicating connection
+logic, and scales well to other connectivity types (e.g.
+``transform=nengo.Sparse(...)``).
+It is also worth noting that it is conceptually similar to the idea of passing
+a ``Distribution`` for the transform, which we are already introducing to
+users.
+
+Note that because we are only modifying the Connection, and not the Ensemble
+parameters, every convolutional neuron will have its own gain and bias.
+This is somewhat unusual, as typically those parameters would be shared within
+each channel.
+However, there is nothing inherently wrong with untieing these parameters
+(it is not uncommonly done), and if this distinction is important to a user
+they can achieve the tied effect by manually setting the gains/biases
+appropriately on the post Ensemble.
```

Comment:
With `nengo_dl`, I assume it's easy to have gains be not-trainable, but have trainable biases? Or have both not trainable?

I guess the one case we'd miss out on is having shared biases across each channel, as you mentioned. As far as I know (both through reading and personal experience), biases are not terribly important, so I don't think we lose too much by not having them trainable (or having them all be trainable).

Reply:
Yeah gains are actually never trainable (or always trainable, if you think of adjusting the weights/encoders as implicitly adjusting gains).  That's because we roll the gains into the encoders (on encoded connections), so there isn't any separable parameter to train there. And it seemed more consistent to have the same set of trainable parameters on neuron-to-neuron connections as well.  Biases are trainable, but that can be turned on or off.

And yeah, I agree that tied/untied biases are not super critical to network behaviour.  I think the main advantage is just (relatively minor) computational savings through a broadcasted bias addition, rather than a functional difference.  And in the case of NengoDL, if someone is really interested in that they still have `TensorNodes` available.